pkg/operator/staticpod/controller/prune/prune_controller.go

```diff
@@ -0,0 +1,223 @@
+	operatorConfigClient common.OperatorClient
+}
+
+func NewPruneController(
+	targetNamespace, staticPodName string,
+	command []string,
+	kubeClient kubernetes.Interface,
+	operatorConfigClient common.OperatorClient,
+	failedRevisionLimit, succeededRevisionLimit int,
+) *PruneController {
+	c := &PruneController{
+		targetNamespace:        targetNamespace,
+		staticPodName:          staticPodName,
+		command:                command,
+		failedRevisionLimit:    failedRevisionLimit,
+		succeededRevisionLimit: succeededRevisionLimit,
+
+		operatorConfigClient: operatorConfigClient,
+		kubeClient:           kubeClient,
+
+		queue:            workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "PruneController"),
+		prunerPodImageFn: getPrunerPodImageFromEnv,
+	}
+
+	return c
+}
+
+func getPrunerPodImageFromEnv() string {
+	return os.Getenv("OPERATOR_IMAGE")
+}
+
+func (c *PruneController) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	glog.Infof("Starting PruneController")
+	defer glog.Infof("Shutting down PruneController")
+
+	// doesn't matter what workers say, only start one.
+	go wait.Until(c.runWorker, time.Second, stopCh)
+
+	<-stopCh
+}
+
+func (c *PruneController) runWorker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *PruneController) processNextWorkItem() bool {
+	dsKey, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(dsKey)
+
+	err := c.sync()
+	if err == nil {
+		c.queue.Forget(dsKey)
+		return true
+	}
+
+	utilruntime.HandleError(fmt.Errorf("%v failed with : %v", dsKey, err))
+	c.queue.AddRateLimited(dsKey)
+
+	return true
+}
+
+func (c *PruneController) sync() error {
+	_, operatorStatus, _, err := c.operatorConfigClient.Get()
+	if err != nil {
+		return err
+	}
+
+	err = c.pruneRevisionHistory(operatorStatus)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func getPrunerPodName(podName, nodeName string) string {
+	return fmt.Sprintf("prune-%s-%s", podName, nodeName)
+}
+
+func (c *PruneController) pruneRevisionHistory(operatorStatus *operatorv1.StaticPodOperatorStatus) error {
+	for _, nodeStatus := range operatorStatus.NodeStatuses {
+		// Read status config map to figure which deployments to keep
+		statusConfigMap, err := c.kubeClient.CoreV1().ConfigMaps(c.targetNamespace).Get(getStatusConfigMapName(nodeStatus.TargetRevision, nodeStatus.NodeName), metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+
+		// store IDs as strings to join for protectedIDs param to pruner pod
+		var succeededRevisionIDs []string
+		var failedRevisionIDs []string
+
+		for revisionID, revisionStatus := range statusConfigMap.Data {
+			switch revisionStatus {
+			case string(corev1.PodSucceeded):
+				succeededRevisionIDs = append(succeededRevisionIDs, revisionID)
+			case string(corev1.PodFailed):
+				failedRevisionIDs = append(failedRevisionIDs, revisionID)
+			}
+		}
+		sort.Strings(succeededRevisionIDs)
```

Comment:
this looks likely to sort improperly.  Add test to expose it.

Reply:
I'm inclined to say we should move the sorting operation into `pruneEligibaleIDs` to get rid of that requirement from callers. You don't care if it is sorted here, either.

Reply:
Don't we use `0001` etc numbering?

Reply:
In that case David's concert about sorting numbers with alpha is critical.